Compfiles/Imo2023P1.lean

```diff
@@ -14,25 +14,445 @@ problem_file { tags := [.NumberTheory] }
+
+  have exp_eq : x_exp + 1 = y_exp := by
+    by_contra ne
+    apply h
+    have lt := Nat.lt_of_le_of_ne exp_lt ne
+
+    exists p ^ (x_exp + 1)
+    refine And.intro ?_ ⟨
+        x_eq_p_exp ▸ Nat.pow_lt_pow_succ p_prime.one_lt,
+        y_eq_p_exp ▸ Nat.pow_lt_pow_of_lt p_prime.one_lt lt
+      ⟩
+
+    let ⟨b, n_eq⟩ := hy
+    let ⟨k, y_eq⟩ := Nat.exists_eq_add_of_lt lt
+    exists b * p ^ (k + 1)
+    rw [n_eq, y_eq_p_exp, y_eq, add_assoc, Nat.pow_add, mul_assoc,
+        mul_comm _ b]
+
+  rw [x_eq_p_exp, y_eq_p_exp, ← exp_eq]
+  exact rfl
+
+/-- Proof that all prime powers are `Dividable`. -/
+lemma PrimePow_is_dividable {n : ℕ} : IsPrimePow n → 1 < n ∧ Dividable n := by
+  intro h
+  refine ⟨h.one_lt, ?_⟩
+
+  obtain ⟨p, e, hp, hpp⟩ := h
+  intro _ _ _ ⟨xy_cons, yz_cons⟩
+  let y_subs := PrimePow_cons_are_p_apart hp.nat_prime ⟨e, hpp⟩ xy_cons
+  let z_subs := PrimePow_cons_are_p_apart hp.nat_prime ⟨e, hpp⟩ yz_cons
+  rw [← z_subs, ← y_subs]
+  exists p + p ^ 2
+  linarith
+
+/-- Proof that all `Dividable` numbers `> 1` are prime powers. -/
+lemma dividable_is_PrimePow {n : ℕ} (h : 1 < n ∧ Dividable n)
+    : IsPrimePow n := by
+  obtain ⟨n_gt_1, hd⟩ := h
+  let p := n.minFac
+  let p_prime := (n.minFac_prime (Nat.ne_of_lt n_gt_1).symm)
+
+  by_contra hn
+  let ⟨q, e, q_prime, q_ne_p, cxy, cyz⟩ := minFac_cons_factor n_gt_1 hn
+  let ⟨x_div_n, y_div_n, _⟩ := cxy
+  let ⟨_, z_div_n, _⟩ := cyz
+  obtain ⟨x, hx⟩ := x_div_n
+  obtain ⟨y, hy⟩ := y_div_n
+  obtain ⟨z, hz⟩ := z_div_n
+
+  have h : p ^ (e + 1) ∣ q * (p + 1) := by
+    let ⟨c, h⟩ :=
+      have n_gt_0 := Nat.zero_lt_of_lt n_gt_1
+      hd ⟨
+        inv_cons_factors n_gt_0 hy hz cyz,
+        inv_cons_factors n_gt_0 hx hy cxy
+      ⟩
+    let h := congrArg (p ^ (e + 1) * q * ·) h
+    exists c
+
+    have : Function.Injective (· * n) :=
+      fun _ _ h => Nat.mul_right_cancel (Nat.zero_lt_of_lt n_gt_1) h
+    apply_fun (· * n)
+    dsimp
+
+    rw [add_comm, mul_add, add_mul, mul_one]
+    nth_rw 1 [hy]
+    nth_rw 2 [hx]
+    nth_rw 3 [hz]
+
+    rw [mul_assoc, ← Nat.mul_add q, ← mul_assoc, ← Nat.pow_add_one',
+        ← Nat.mul_add, ← mul_assoc, mul_comm q, mul_assoc _ c, mul_rotate' c,
+        ← mul_assoc]
+    exact h
+
+  have : 0 < (p ^ (e + 1)).factorization p := by
+    rw [p_prime.factorization_pow, Finsupp.single_eq_same]
+    exact Nat.zero_lt_succ _
+  apply Nat.not_le_of_gt this
+
+  -- Show that `q * (p + 1)` is not divisible by `p`.
+  have : (q * (p + 1)).factorization p = 0 := by
+    rw [Nat.factorization_mul q_prime.ne_zero p.succ_ne_zero]
+    dsimp
+    rw [q_prime.factorization, Finsupp.single_eq_of_ne q_ne_p,
+        p_succ_fact_zero p_prime]
+  refine this ▸ (Nat.factorization_le_iff_dvd ?_ ?_).mpr h p
+  . exact (pow_ne_zero (e + 1) p_prime.ne_zero)
+  . exact (Nat.mul_ne_zero q_prime.ne_zero p.succ_ne_zero)
+
+snip end
+
+determine solution_set : Set ℕ := { n | IsPrimePow n }
+
+problem imo2023_p1 : solution_set = { n | 1 < n ∧ Dividable n } := by
```

Comment:
Joseph Myers pointed out [on the Lean Zulip](https://leanprover.zulipchat.com/#narrow/channel/208328-IMO-grand-challenge/topic/Suggested.20formalization.20conventions/near/519250109) that this new formal statement of the problem dropped the requirement that `n` be composite. 

I added that back here: https://github.com/dwrensha/compfiles/commit/38b3b2680621d7fe45e6634c8c1be72cd8f80a06

The aim is to make the formal problem statements match the informal statements as closely as possible.